### PR TITLE
Updated UTF to spin up a cluster, interrogate it, then spin it down

### PR DESCRIPTION
**Pull request by chelma**
### Description
* Added library code to manage Docker interactions
* Added library code to spin up and manage an ElasticSearch Cluster and its Nodes
* Added unit tests for the library code
* Updated the UTF workflow to spin up a cluster of 3 nodes, query the cluster state, then spin down the cluster
* Updated the `run_utf.py` script to kick off the workflow
* Deprecated the "resume" functionality as it was becoming cumbersome to keep up-to-date and seems to offer limited value
* Bare-bones `test_config.json` parsing

### Issues Resolved
* https://github.com/opensearch-project/opensearch-migrations/issues/6
* https://github.com/opensearch-project/opensearch-migrations/issues/8

### Testing
Added unit tests:
```
(.venv) chelma@3c22fba4e266 upgrades % python -m pytest tests/
================================================================================ test session starts =================================================================================
platform darwin -- Python 3.10.6, pytest-7.2.0, pluggy-1.0.0
rootdir: /Users/chelma/workspace/OSMigration/opensearch-migrations/upgrades
collected 56 items                                                                                                                                                                   

tests/test_cluster.py ......                                                                                                                                                   [ 10%]
tests/test_docker_framework_client.py .............                                                                                                                            [ 33%]
tests/test_exception_base.py ...                                                                                                                                               [ 39%]
tests/test_framework_runner.py ......                                                                                                                                          [ 50%]
tests/test_framework_state.py ...                                                                                                                                              [ 55%]
tests/test_framework_step.py ....                                                                                                                                              [ 62%]
tests/test_logging_wrangler.py .                                                                                                                                               [ 64%]
tests/test_node.py .......                                                                                                                                                     [ 76%]
tests/test_shell_interactions.py .......                                                                                                                                       [ 89%]
tests/test_test_config_wrangling.py .....                                                                                                                                      [ 98%]
tests/test_workspace_wrangler.py .                                                                                                                                             [100%]

================================================================================= 56 passed in 1.79s =================================================================================
```

Also, tested workflow using the UTF kickoff script.  My `test_config.json` looks like this:
```
{
    "source_docker_image": "docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2"
}

```

The script and its output looks like this:
```
(.venv) chelma@3c22fba4e266 upgrades % ./run_utf.py --test_config /tmp/test_config.json
[FrameworkRunner] Running Step: LoadTestConfig
[LoadTestConfig] Loading test config file...
[LoadTestConfig] Loaded test config file successfully
[FrameworkRunner] Step Succeeded: LoadTestConfig
[FrameworkRunner] Running Step: BootstrapDocker
[BootstrapDocker] Checking if Docker is installed and available...
[BootstrapDocker] Docker appears to be installed and available
[BootstrapDocker] Ensuring the Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available either locally or remotely...
[BootstrapDocker] Docker image docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2 is available
[FrameworkRunner] Step Succeeded: BootstrapDocker
[FrameworkRunner] Running Step: StartSourceCluster
[StartSourceCluster] Creating source cluster...
[StartSourceCluster] Waiting up to 30 sec for cluster to be active...
Node source-cluster-node-1 is now active
Node source-cluster-node-3 is now active
Node source-cluster-node-2 is now active
[StartSourceCluster] Cluster source-cluster is active
[FrameworkRunner] Step Succeeded: StartSourceCluster
[FrameworkRunner] Running Step: TestSourceCluster
[TestSourceCluster] ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role master name
172.25.0.2           31          59  68    2.15    0.51     0.24 dimr      -      source-cluster-node-1
172.25.0.4           10          59  58    2.15    0.51     0.24 dimr      -      source-cluster-node-3
172.25.0.3           12          59  58    2.15    0.51     0.24 dimr      *      source-cluster-node-2
[FrameworkRunner] Step Succeeded: TestSourceCluster
[FrameworkRunner] Running Step: StopSourceCluster
[StopSourceCluster] Stopping cluster source-cluster...
[StopSourceCluster] Cleaning up underlying resources for cluster source-cluster...
[FrameworkRunner] Step Succeeded: StopSourceCluster
[FrameworkRunner] Ran through all steps successfully
[FrameworkRunner] Saving application state to file...
[FrameworkRunner] Application state saved
[FrameworkRunner] Application state saved to: /tmp/utf/state-file
[FrameworkRunner] Full run details logged to: /tmp/utf/logs/run.log.2022-12-12_12_28_20
```

### Check List
- [X] New functionality includes testing
  - [X] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: upgrades/upgrade_testing_framework/steps/step_load_test_config.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        # Begin the step body
+        self.logger.info("Loading test config file...")
+        try:
+            test_config = tcw.load_test_config(test_config_path)
+        except Exception as exception:
+            self.fail(f"Unable to load test config file.  Details: {str(exception)}", exception)
+        self.logger.info("Loaded test config file successfully")
+        
+        # Update our state
+        self._set_state_value("test_config", test_config)
```
Comment (mikaylathompson) [inline upgrades/upgrade_testing_framework/steps/step_load_test_config.py:23]: No issues, but points for discussion:
1. What's your thinking between putting the test_config as a full object in the state vs. unpacking it to multiple (in the future) top-level items?
2. Do you think of the state as read/append-only?  This is a case where it feels not-unreasonable to replace the file path with the config itself, but I'm I think I have been overall assuming it's append-only -- could be worth formalizing our expectations for how steps interact with the state.

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        self.logger.info("Creating source cluster...")
+        cluster = Cluster("source-cluster", 3, docker_client, source_docker_image)
+        cluster.start()
+
+        try: 
+            max_wait_sec = 30
+            self.logger.info(f"Waiting up to {max_wait_sec} sec for cluster to be active...")
+            cluster.wait_for_cluster_to_start_up(wait_limit_sec=30)
+        except ClusterNotStartedInTimeException as exception:
+            self.fail("Cluster did not start within time limit", exception)
```
Comment (mikaylathompson) [inline upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:24]: should anything be saved to the state in this case? we possibly have a cluster that was created but not started--should there be a record of it in the state?

File: upgrades/upgrade_testing_framework/steps/step_test_source_cluster.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    """
+    This step is where you run tests on the source cluster
+    """
+
+    def _run(self):
+        # Get the state we need
+        # N/A
+
+        # Begin the step body
+        _, output = shell.call_shell_command("curl -X GET \"localhost:9200/_cat/nodes?v=true&pretty\"")
```
Comment (mikaylathompson) [inline upgrades/upgrade_testing_framework/steps/step_test_source_cluster.py:16]: a just curious question: any particular reason to do this via shell command vs. opensearch-py library? (thought is that the python client has a fair amount of error handling/possibly retries built in already)

(actual change would be out of scope of this PR, but for the future)

File: upgrades/tests/test_cluster.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
+import pytest
+import unittest.mock as mock
+
+import upgrade_testing_framework.cluster_management.cluster as cluster
+
+def test_WHEN_create_Cluster_AND_not_enough_nodes_THEN_raises():
+    # Set up test
+    test_cluster_name = "cluster-name"
+    test_num_nodes = 0 # must be > 1
```
Comment (mikaylathompson) [inline upgrades/tests/test_cluster.py:9]: Minor point: `# must be >= 1`

at least according to the logic here: https://github.com/opensearch-project/opensearch-migrations/pull/40/files#diff-4409f008baff38f77bc0c7eb477861e6df14bf9abcac1a01355a68922ce0c68eR37

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                if self._nodes[node_name].is_active():
+                    self.logger.info(f"Node {node_name} is now active")
+                    nodes_to_wait_for.remove(node_name)
+
+            if not nodes_to_wait_for: # all nodes active
+                break
+            
+            # TODO: Check resolution is wait_interval_sec, so might not respect wait_limit_sec if > 1
+            if not None == wait_limit_sec and total_wait_time_sec >= wait_limit_sec:
+                raise ClusterNotStartedInTimeException()
```
Comment (mikaylathompson) [inline upgrades/upgrade_testing_framework/cluster_management/cluster.py:108]: Another thought here -- we need to deal with cleanup. presumably some of the nodes may have started successfully and need to be stopped, and all of them should be cleaned-up.

I'm fine with it being mostly out-of-scope, but maybe helpful to at least log a list of which ones did start and need to be stopped?

File: upgrades/tests/test_cluster.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    mock_docker_client = mock.Mock()
+    test_image = "image"
+
+    # Run our test
+    with pytest.raises(ValueError):
+        cluster.Cluster(test_cluster_name, test_num_nodes, mock_docker_client, test_image)
+
```
Comment (mikaylathompson) [inline upgrades/tests/test_cluster.py:16]: nitpick: I suspect flake8 is going to suggest that there should be 2 lines between each of these (https://peps.python.org/pep-0008/#blank-lines) -- maybe try running the auto-transform thing on this? Or just wait for it to complain 🙃

File: upgrades/tests/test_test_config_wrangling.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+TEST_CONFIG = {"source_docker_image": "my_image"}
+
+@pytest.fixture
+def valid_test_config_file(tmpdir):
+    test_config_file = py.path.local(os.path.join(tmpdir.strpath, "test_config.json"))
+    test_config_file.write(json.dumps(TEST_CONFIG, sort_keys=True, indent=4))
+    return test_config_file
+
+@pytest.fixture
+def not_parsible_file(tmpdir):
```
Comment (mikaylathompson) [inline upgrades/tests/test_test_config_wrangling.py:17]: nitpick: parsible -> parsable (or parseable) [[source](https://en.wiktionary.org/wiki/parsable)]

File: upgrades/tests/test_test_config_wrangling.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    # Run our test
+    actual_value = tcw.load_test_config(valid_test_config_file.strpath)
+
+    # Check the results
+    expected_value = TEST_CONFIG
+    assert expected_value == actual_value
+
+def test_WHEN_load_test_config_called_AND_doesnt_exist_THEN_raises(tmpdir):
+    # Set up our test
+    non_existant_file = os.path.join(tmpdir.strpath, "test_config.json")
```
Comment (mikaylathompson) [inline upgrades/tests/test_test_config_wrangling.py:38]: nitpick: existant -> existent

**Review by mikaylathompson — COMMENTED** [review]
High level: I love it. I think the docker framework + higher level constructs look like a really good for-now solution, and it's all quite readable.

I have a combo of minor suggestions, spelling and formatting nitpicks, and open-ended questions.

File: upgrades/tests/test_cluster.py (hunk 4 above)
Comment (chelma) [reply upgrades/tests/test_cluster.py:9]: Ah, good callout

File: upgrades/tests/test_test_config_wrangling.py (hunk 7 above)
Comment (chelma) [reply upgrades/tests/test_test_config_wrangling.py:17]: Thanks, will fix

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    def __init__(self):
+        super().__init__(f"The Cluster did not start up within the specified time frame")
+
+"""
+The goal of this class's abstraction is to isolate the rest of the codebase from the details of precisely how we
+are implementing clusters for our tests.  Ideally, the rest of the code should not know of things like where the
+nodes are hosted (e.g. local Docker, local installs, etc).  The current implementation fell a bit short of that, but
+we can iterate.
+
+Additionally, Since the initial cut is focused on upgrades via the snapshot/restore mechanism, it feels like we can
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:21]: Thx

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        self._volumes: List[Volume] = []
+
+    def _generate_network_name(self):
+        return f"{self.name}-network"
+
+    def _generate_node_name(self, number: int):
+        return f"{self.name}-node-{number}"
+
+    def _generate_volume_name(self, suffix: str):
+        return f"{self.name}-volume-{suffix}"
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:49]: Yeah, they should.  Thanks!

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        for node_num in range(1, self._num_starting_nodes + 1):
+            # Create the node's data volume
+            node_volume = self._docker_client.create_volume(self._generate_volume_name(node_num))
+            self._volumes.append(node_volume)
+
+            # Create the node name, node config, container config
+            node_name = self._generate_node_name(node_num)
+            node_config = NodeConfiguration(node_name, self.name, starting_master_nodes, starting_seed_hosts)
+            port_mappings = [
+                PortMapping(f"920{node_num - 1}", "9200") # TODO: Breaks if more than 10 nodes (unlikely?)
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:74]: I already have a fix for this in my changes to address the next task (https://github.com/opensearch-project/opensearch-migrations/issues/7).  I change how the port stuff works.  Mind leaving this one for now?

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 5 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:108]: All of the cluster state edge cases need more work.  I'm planning on iterating on those in the next pass (https://github.com/opensearch-project/opensearch-migrations/issues/7).  Mind waiting until the PR for that drops before looking at stuff like this too closely?

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            time.sleep(wait_interval_sec)
+
+    def stop(self):
+        # TODO: handle when the container is not running
+        
+        self.logger.debug(f"Stopping cluster {self.name}...")
+
+        # call stop() on each of the Nodes
+        for node in self._nodes.values():
+            node.stop()
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:121]: I agree, it needs more work.  I'm planning on iterating on those in the next pass (https://github.com/opensearch-project/opensearch-migrations/issues/7).  Mind waiting until the PR for that drops before looking at stuff like this too closely?

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+        # TODO: confirm all nodes are stopped(?)
+
+        self.logger.debug(f"Stopped cluster {self.name}")
+
+    def clean_up(self):
+        # TODO: handle when the cluster is not stopped
+
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:129]: Thx

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        # remove network
+        for network in self._networks:
+            self._docker_client.remove_network(network)
+        
+        self.logger.debug(f"Cleaned up cluster {self.name}")
+            
+
+    
+        
+
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:150]: Ooof, don't remind me!

File: upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+            image,
+            name=container_name,
+            network=network.name,
+            ports=port_mapping,
+            volumes=volume_mapping,
+            ulimits=ulimits,
+            detach=True,
+            environment=env_variables
+        )
+        self.logger.debug("Created container")
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py:119]: Yep, makes sense.  Thanks!

File: upgrades/upgrade_testing_framework/core/framework_state.py [hunk 16]
```diff
@@ -1,30 +1,33 @@
...
+for how the rest of the code should be interacting with it.
+
+Additionally, want to be able to dump all useful information to the state file for debugging purposes.  However, we
+currently only dump the key/value pairs in the _state_dict.  As a result, we don't get any information about the
+cluster we're storing as an internal member during the FrameworkSteps.  We need to invest in this area of the code.
+"""
 class FrameworkState:
     def __init__(self, state: dict):
-        self._state = state
+        self.docker_client: DockerFrameworkClient | None = None
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/core/framework_state.py:16]: I'm learning type hints for the first time too but I think they're the same.  I agree that optional is better.

File: upgrades/upgrade_testing_framework/core/test_config_wrangling.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
+import json
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/core/test_config_wrangling.py:1]: Yeah, you point pytest at a specific directory tree and this isn't in it.  Also, pytest isn't imported in the file which I think plays a factor as well.

File: upgrades/upgrade_testing_framework/steps/step_load_test_config.py (hunk 1 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_load_test_config.py:23]: Good questions.

1. I don't have strong opinions about this at the moment; as long as its easy to write/maintain, I'm down for whatever.  That said, it feels a bit better to have separate objects for each highest-level `test_config.json` key just to make the individual unit tests simpler.
2. Yeah, I'm thinking about it as append-only.  That would be good to call out somewhere explicitly, per your comment.

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py (hunk 2 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:24]: Good callout.  I feel like we need to have something like "cleanup" steps that are run in the equivalent of a `finally` block in the FrameworkRunner's main loop to ensure cleanup happens.  I'll think on that.

In this particular case, there isn't any value in saving the cluster to the state because the FrameworkState is a mess and most stuff isn't actually serialized to the `state-file`, so... maybe not?  I can see an argument to add that now so we don't forget to handle the edge case later though, even if there's no payoff now.

File: upgrades/upgrade_testing_framework/steps/step_test_source_cluster.py (hunk 3 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_test_source_cluster.py:16]: No particular reason than this took me ~1 min to write rather than ~30 minutes.  Pulling the OpenSearch SDK into the Framework is something we should do eventually.

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:49]: Oh - I was trying to figure out why they weren't there, and discovered that I fixed them in the changes I made for the follow-up task, after I posted this PR.

File: upgrades/upgrade_testing_framework/cluster_management/node.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        self._docker_client.remove_container(self._container)
+        self.logger.debug(f"Container {self._container.name} has been deleted")
+        self._container = None
+
+    def is_active(self):
+        self.logger.debug(f"Checking if node {self.name} is active...")
+        if self._container == None:
+            return False
+        
+        exit_code, output = self._docker_client.run(self._container, "curl -X GET \"localhost:9200/\"")
```
Comment (lewijacn) [inline upgrades/upgrade_testing_framework/cluster_management/node.py:90]: Any particular reason to make a curl request here (Which may be more malleable in future with say security plugin enabled or a different port configured) then say doing something like this: https://tech.serhatteker.com/post/2021-04/check-docker-running-with-python/ or fetching the list of running containers and see if this one exists

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    def _run(self):
+        # Get the state we need
+        source_docker_image = self._get_state_value("test_config")["source_docker_image"]
+        docker_client = self.state.docker_client
+
+        # Begin the step body
+        self.logger.info("Creating source cluster...")
+        cluster = Cluster("source-cluster", 3, docker_client, source_docker_image)
```
Comment (lewijacn) [inline upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:16]: There's a lot of configuration we could be passing in here from the test configuration, I am assuming we just haven't got to that point yet but will in the future

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        cluster.start()
+
+        try: 
+            max_wait_sec = 30
+            self.logger.info(f"Waiting up to {max_wait_sec} sec for cluster to be active...")
+            cluster.wait_for_cluster_to_start_up(wait_limit_sec=30)
+        except ClusterNotStartedInTimeException as exception:
```
Comment (lewijacn) [inline upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:23]: What are your thoughts on cleaning up created resources on failed start ups? Maybe this is a flag we can add in the configuration that some users may want and some may not

File: upgrades/tests/test_cluster.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    test_cluster.clean_up()
+
+    # Check the results
+    assert 1 == mock_node_1.clean_up.call_count
+    assert 1 == mock_node_2.clean_up.call_count
+
+    expected_remove_networks = [mock.call(mock_network)]
+    assert expected_remove_networks == test_cluster._docker_client.remove_network.call_args_list
+
+    expected_remove_networks = [mock.call(mock_volume_1), mock.call(mock_volume_2)]
```
Comment (lewijacn) [inline upgrades/tests/test_cluster.py:161]: I think this should be `expected_remove_volumes`

File: upgrades/tests/test_node.py [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+        "cluster.name": test_cluster_name,
+        "cluster.initial_master_nodes": ",".join(test_master_nodes),
+        "discovery.seed_hosts": ",".join(test_seed_hosts),
+        "node.name": test_node_name,
+        "bootstrap.memory_lock": "true",
+        "ES_JAVA_OPTS": "-Xms512m -Xmx512m"
+    }
+    assert expected_value == actual_value.config
+
+def test_WHEN_node_start_THEN_as_expected():
```
Comment (lewijacn) [inline upgrades/tests/test_node.py:27]: A lot of these cases are happy cases, which I am assuming is temporary and will change once we start tackling some edge cases and add to this

**Review by lewijacn — COMMENTED** [review]
Structuring of this is very nice. I had some minor comments but am working under the assumption that we will focus on more of the edge cases in later revisions

File: upgrades/tests/test_cluster.py (hunk 21 above)
Comment (chelma) [reply upgrades/tests/test_cluster.py:161]: Ah, yep - thanks!

File: upgrades/tests/test_node.py (hunk 22 above)
Comment (chelma) [reply upgrades/tests/test_node.py:27]: Yeah, that's the plan.  I want to do that in either the next revision (in-progress) or the one after.  TBH though, trying to work backwards from need so it may just happen the first time one of us runs into a problem with it.

File: upgrades/tests/test_test_config_wrangling.py (hunk 8 above)
Comment (chelma) [reply upgrades/tests/test_test_config_wrangling.py:38]: Thx

File: upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+        volume = self._docker_client.volumes.create(name)
+        self.logger.debug(f"Created volume {name}")
+        return volume
+
+    def remove_volume(self, volume: Volume):
+        self.logger.debug(f"Removing volume {volume.name}...")
+        volume.remove()
+        self.logger.debug(f"Removed volume {volume.name}")
+
+    def create_container(self, image: str, container_name: str, network: Network, ports: List[PortMapping], volumes: List[Volume], ulimits: List[Ulimit],
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py:97]: Thanks!  The Docker SDK is probably better to use than roll our own, but it certainly has its issues.

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py (hunk 19 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:16]: Yeah, I'm fleshing that out right now in my code for the next task.

File: upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py (hunk 20 above)
Comment (chelma) [reply upgrades/upgrade_testing_framework/steps/step_start_source_cluster.py:23]: Nice-to-have, but not required-right-now.  I talk about this a bit in my response to Mikayla a few lines down.  TBH, there's a lot of nice-to-haves that I'm deliberately postponing so that we can iterate quickly to hit our Demo goals.

File: upgrades/tests/test_cluster.py (hunk 4 above)
Comment (chelma) [reply upgrades/tests/test_cluster.py:9]: Addressed in next rev

File: upgrades/tests/test_cluster.py (hunk 21 above)
Comment (chelma) [reply upgrades/tests/test_cluster.py:161]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:21]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:49]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+            for node_name in nodes_to_wait_for:
+                if self._nodes[node_name].is_active():
+                    self.logger.info(f"Node {node_name} is now active")
+                    nodes_to_wait_for.remove(node_name)
+
+            if not nodes_to_wait_for: # all nodes active
+                break
+            
+            # TODO: Check resolution is wait_interval_sec, so might not respect wait_limit_sec if > 1
+            if not None == wait_limit_sec and total_wait_time_sec >= wait_limit_sec:
```
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:107]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:129]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/cluster.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/cluster.py:150]: Addressed in next rev

File: upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/cluster_management/docker_framework_client.py:119]: Addressed in next rev

File: upgrades/upgrade_testing_framework/core/framework_state.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chelma) [reply upgrades/upgrade_testing_framework/core/framework_state.py:16]: Addressed in next rev

**Review by mikaylathompson — APPROVED** [review]
Via discussion off-gh, it's not possible for workflow reasons to push another commit to this PR. Seeing as none of these comments are blocking, I'm approving this to be merged now, with a follow up PR that addresses the comments.

**Review by lewijacn — APPROVED** [review]
